docs/life_pt1.md

```diff
@@ -0,0 +1,293 @@
+
+# 13.1 Well...
+
+> "No." - _Doom Slayer_
+
+Well, someone can, probably. But doom in a dataframe would be kinda hard to play, so let's try something simpler.
+[Conway's Game of Life](https://en.wikipedia.org/wiki/Conway%27s_Game_of_Life) is a notorious Cellular Automaton that we could perhaps implement with a plugin.
+For science, of course.
+
+![Toad pattern with period = 2](assets/life_toad.gif)
+
+Jokes aside, life allows us to show how a plugin can access elements in both neighbouring rows and columns for each element.
+With a little bit of extra Python, we can display things in an almost pretty manner.
+
+> __Note:__ For this tutorial, we'll assume you created a new plugin from the cookiecutter template and named it `game_of_life` (these steps aren't shown here, since they were already covered at the very beginning of this series).
+
+In this section we'll cover the developer side of the plugin (both Python and Rust).
+In the next section we'll show how a user can import and use what we developed here.
+
+## The Python side
+
+Let's take a look at what we'll implement first, in `game_of_life/__init__.py`:
+
+```python
+import fileinput
+from collections import OrderedDict
+from itertools import tee, islice
+from os import PathLike
+from pathlib import Path
+from typing import Iterable, Any
+
+import polars as pl
+from polars._typing import IntoExpr
+
+from game_of_life.utils import register_plugin
+
+
+# Parse a board from a file or stdin
+def parse_board(ifile: str | ...) -> list[list[int]]: ...
+
+# Transpose a list of lists
+def _transpose(board: list[list[int]]) -> list[list[int]]: ...
+
+# Creates a DataFrame from a list of lists
+def board_to_df(board: list[list[int]]) -> pl.DataFrame: ...
+
+# Helper function to help us deal with corner cases
+def _nwise_wrapping(iterable: Iterable[Any], n: int): ...
+
+# Advance the simulation by n steps
+def step(df: pl.DataFrame, n: int = 1): ...
+
+# Register our plugin
+def life_step(left: IntoExpr, mid: IntoExpr, right: IntoExpr) -> pl.Expr: ...
+```
+
+Starting with the function to parse a board from a file or stdin:
+
+```python
+def parse_board(
+        ifile: str
+               | bytes
+               | PathLike[str]
+               | PathLike[bytes]
+               | Iterable[str | bytes | PathLike[str] | PathLike[bytes]],
+) -> list[list[int]]:
```

Comment:
shall we add an autoformetter to standardise on formatting here?

see here for an example: https://github.com/narwhals-dev/narwhals/blob/cb4a583b9b83c23687c09bbe0ad55ab43d790de9/.pre-commit-config.yaml#L42-L48

Reply:
TBH this function didn't even need to be that... pedantic (I think I formatted it this way to avoid a huge line).
I like the autoformatter, but this is in (partial) conflict with the idea of moving the code snippets into separate, dedicated files.
In any case, I'd rather do this in a separate PR, is that ok?